File: rice_cooker.go
package main

import (
	"fmt"
	"os"
	"strconv"
)

var (
	powerConnected bool
	lidOpen        bool
	foodAdded      bool
	waterAdded     bool
	lidClosed      bool
	cooking        bool
	maxCupLimit    int
	foodQuantity   int
	waterQuantity  int
)

func main() {
	for {
		displayMenu()
		var choice int
		fmt.Print("Choix : ")
		_, err := fmt.Scan(&choice)
		if err != nil {
			fmt.Println("Une erreur s'est produite :", err)
			os.Exit(1)
		}
		handleUserChoice(choice)
	}
}

func displayMenu() {
	fmt.Println("Voici les étapes à suivre :")
	fmt.Println("1. Brancher la prise")
	fmt.Println("2. Ouvrir le couvercle")
	fmt.Println("3. Verser l'aliment")
	fmt.Println("4. Verser de l'eau")
	fmt.Println("5. Fermer le couvercle")
	fmt.Println("6. Mettre le bouton sur cook")
	fmt.Println("7. Faire cuire")
	fmt.Println("8. Débrancher la prise")
	fmt.Println("0. Quitter")
}

func handleUserChoice(choice int) {
	switch choice {
	case 1:
		connectPower()
	case 2:
		openLid()
	case 3:
		pourFood()
	case 4:
		pourWater()
	case 5:
		closeLid()
	case 6:
		startCooking()
	case 7:
		cook()
	case 8:
		disconnectPower()
	case 0:
		fmt.Println("Programme terminé.")
		os.Exit(0)
	default:
		fmt.Println("Option invalide. Veuillez choisir à nouveau.")
	}
}

func connectPower() {
	if !powerConnected {
		powerConnected = true
		fmt.Println("La prise est branchée, le courant passe")
		fmt.Println("Le bouton warm s'allume")
	} else {
		fmt.Println("La prise est déjà branchée")
	}
}

func openLid() {
	if !lidOpen {
		lidOpen = true
		fmt.Println("Le couvercle est ouvert")
	} else {
		fmt.Println("Le couvercle est déjà ouvert")
	}
}

func pourFood() {
	if lidOpen {
		fmt.Print("Quantité d'aliment (en tasses) : ")
		var input string
		fmt.Scan(&input)
		foodQuantity, err := strconv.Atoi(input)
		if err != nil {
			fmt.Println("Erreur lors de la saisie de la quantité d'aliment :", err)
			return
		}

		if foodQuantity > 0 && foodQuantity <= maxCupLimit {
			foodAdded = true
			fmt.Println("Aliment ajouté avec succès")
		} else {
			fmt.Println("Quantité invalide d'aliment")
		}
	} else {
		fmt.Println("Le couvercle doit être ouvert pour verser l'aliment")
	}
}

func pourWater() {
	if lidOpen && foodAdded {
		fmt.Print("Quantité d'eau par tasses : ")
		var input string
		fmt.Scan(&input)
		waterQuantity, err := strconv.Atoi(input)
		if err != nil {
			fmt.Println("Erreur lors de la saisie de la quantité d'eau :", err)
			return
		}

		if waterQuantity > 0 && waterQuantity <= maxCupLimit {
			waterAdded = true
			fmt.Println("Eau ajoutée avec succès")
		} else {
			fmt.Println("Quantité invalide d'eau")
		}
	} else {
		fmt.Println("Le couvercle doit être ouvert et l'aliment doit être ajouté avant de verser de l'eau")
	}
}

func closeLid() {
	if lidOpen {
		lidClosed = true
		fmt.Println("Le couvercle est fermé")
	} else {
		fmt.Println("Le couvercle est déjà fermé")
	}
}

func startCooking() {
	if lidClosed {
		cooking = true
		fmt.Println("La cuisson a commencé")
	} else {
		fmt.Println("Le couvercle doit être fermé pour démarrer la cuisson")
	}
}

func cook() {
	if cooking && waterQuantity >= foodQuantity {
		fmt.Println("L'aliment est cuit")
		switchToWarm()
	} else {
		fmt.Println("La cuisson n'est pas encore prête")
	}
}

func switchToWarm() {
	if cooking {
		cooking = false
		fmt.Println("Le bouton est passé sur warm")
	} else {
		fmt.Println("La cuisson doit être terminée pour passer sur warm")
	}
}

func disconnectPower() {
	if powerConnected {
		powerConnected = false
		fmt.Println("Le courant ne passe plus")
	} else {
		fmt.Println("Le rice cooker est déjà débranché")
	}
}

File: rice_cooker.rb
class RiceCooker
    def initialize
      @power_connected = false
      @lid_open = false
      @food_added = false
      @water_added = false
      @lid_closed = false
      @cooking = false
      @max_cup_limit = 5
      @food_quantity = 0
      @water_quantity = 0
    end
  
    def display_menu
      puts "Here are the steps to follow:"
      puts "1. Connect the power"
      puts "2. Open the lid"
      puts "3. Pour the food"
      puts "4. Pour water"
      puts "5. Close the lid"
      puts "6. Set the button to cook"
      puts "7. Cook"
      puts "8. Disconnect the power"
      puts "0. Quit"
    end
  
    def handle_user_choice(choice)
      case choice
      when 1 then connect_power
      when 2 then open_lid
      when 3 then pour_food
      when 4 then pour_water
      when 5 then close_lid
      when 6 then start_cooking
      when 7 then cook
      when 8 then disconnect_power
      when 0 then exit_program
      else
        puts "Invalid option. Please choose again."
      end
    end
  
    def connect_power
      if !@power_connected
        @power_connected = true
        puts "The power is connected, the current is flowing"
        puts "The warm button lights up"
      else
        puts "The power is already connected"
      end
    end
  
    def open_lid
      if !@lid_open
        @lid_open = true
        puts "The lid is open"
      else
        puts "The lid is already open"
      end
    end
  
    def pour_food
      if @lid_open
        puts "Quantity of food (in cups): "
        food_quantity = gets.to_i
        if food_quantity > 0 && food_quantity <= @max_cup_limit
          @food_added = true
          @food_quantity += food_quantity
          puts "Food added successfully"
        else
          puts "Invalid quantity of food"
        end
      else
        puts "The lid must be open to pour the food"
      end
    end
  
    def pour_water
      if @lid_open && @food_added
        puts "Quantity of water (in cups): "
        water_quantity = gets.to_i
        if water_quantity > 0 && water_quantity <= @max_cup_limit
          @water_added = true
          @water_quantity += water_quantity
          puts "Water added successfully"
        else
          puts "Invalid quantity of water"
        end
      else
        puts "The lid must be open, and food must be added before pouring water"
      end
    end
  
    def close_lid
      if @lid_open
        @lid_closed = true
        puts "The lid is closed"
      else
        puts "The lid is already closed"
      end
    end
  
    def start_cooking
      if @lid_closed
        @cooking = true
        puts "Cooking has started"
      else
        puts "The lid must be closed to start cooking"
      end
    end
  
    def cook
      if @cooking && @water_quantity >= @food_quantity
        puts "The food is cooked"
        switch_to_warm
      else
        puts "Cooking is not ready yet"
      end
    end
  
    def switch_to_warm
      if @cooking
        @cooking = false
        puts "The button has switched to warm"
      else
        puts "Cooking must be finished to switch to warm"
      end
    end
  
    def disconnect_power
      if @power_connected
        @power_connected = false
        puts "The current is no longer flowing"
      else
        puts "The rice cooker is already disconnected"
      end
    end
  
    def exit_program
      puts "Program terminated."
      exit
    end
  end
  
  rice_cooker = RiceCooker.new
  
  begin
    while true
      rice_cooker.display_menu
      print "Choice: "
      choice = gets.to_i
      rice_cooker.handle_user_choice(choice)
    end
  rescue => e
    puts "An error occurred: #{e.message}"
  end
  